[all] Add a new `Concept` component, and implement the partials folder structure for roadmap+concepts

heysokam opened the pull request:
> _Note: Contains commits from the `new-MetaStructure` branch_
> _#150 hasn't been merged yet, and this PR builds on it_
> _As such, completing/merging #150 is a hard dependency before merging this PR_
- [x] Depends on Merging: #150
- [x] Create the Concepts `_partials/*` skeleton structure
  - [x] Typescript
  - [x] Visualscript
  - [x] Creator
- [x] Create the `Concept` custom Component
  - [x] Support for Concepts
    - [x] Creation
    - [x] Automated Storage
    - [x] Cross-Referencing stored concepts
  - [x] Create CSS styling for `onHover` concept + tooltip/card rendering
  - [x] Create CSS styling for `block` (aka paragraph) rendering
  - [x] Fix multi-line definitions (children) forcefully adding an extra `</br>` at the end.
  - [x] Fix `.kind=` property access error triggering when clicking on a `#` ToC link
  - [x] ~~Attempt to add a `.hLevel={Number}` property to the component _(likely requires v3.2.1)_~~
- [x] Add a `<Concept />` component with dummy description+title+category for every concept
  - [x] Typescript
    - [x] Add the folder of the file to the description of the typescript concepts
  - [x] Visualscript
  - [x] Creator

### QA Notes
 @armando-salazar Review:
 - [x] `_partials/concepts/*` folder structure
 - [x] `_partials/roadmap/*` folder structure
 - [x] Roadmap:
   - [x] Introductory text: first draft
   - [x] Each Learning Site Roadmap
     - [x] Creator
     - [x] Visualscript
     - [x] Typescript 
 
 @DanielBelmes Review:
 - [x] Concepts Component
   - [x] Source Code
   - [x] Storage "backend"  (internal global-state with a TS `Map<string,Concept>`)
   - [x] Feature set
   - [x] UX

DanielBelmes commented on src/components/Concept.tsx:
```diff
@@ -0,0 +1,256 @@
+import Error from '@theme/Error';
```
delete this

heysokam replied:
Resolved with [08842b0](https://github.com/EtherealEngine/etherealengine-docs/pull/156/commits/08842b05237dae37e4c9b8ab7d754e07b4a0da77)

armando-salazar commented:
@heysokam, to enhance accessibility and improve navigation within the extensive list of concepts in the Roadmap, I recommend improving the usage of the table of contents. This would allow users to navigate the document more easily and locate concepts quickly. Could we modify the partials to use a level 3 header (###) for the concepts? 

This adjustment would make the ToC more effective as a navigation tool, helping users to more easily view and find information.

heysokam commented:
> @heysokam, to enhance accessibility and improve navigation within the extensive list of concepts in the Roadmap, I recommend improving the usage of the table of contents. This would allow users to navigate the document more easily and locate concepts quickly. Could we modify the partials to use a level 3 header (`###`) for the concepts?
>
> This adjustment would make the ToC more effective as a navigation tool, helping users to more easily view and find information.

@armando-salazar This change to the Concepts custom component would break every future usage of the component.
https://tsu.atlassian.net/wiki/spaces/IREE/pages/2181955616/Markdown+Partials#Rules-and-Guidelines
Partials (and the same goes for Components, as they are inserted in the same way through MDX) cannot contain `#` (or any other level) titles, or they will break the Table of Contents of the pages they are introduced into.

armando-salazar commented:
After reviewing, I've identified several key points:

1. The `partials` imported into the Roadmap from the Creators and VisualScript sites use `title`s that merely indicate a relative path, such as **`assets/cms/panel/description`** from Creators, in contrast to TypeScript which uses a descriptive custom `title` like `Logic vs Data`.

2. It appears that these concepts are intended to be categorized, as indicated by labels such as "Advanced Non-essentials", "Mastery", etc. Is there a directory organization within the system that reflects these categories?

3. Regarding the categories mentioned in point 2, it could be beneficial to define what each category encompasses. For example, the term "Advanced Non-essentials" could be confusing— isn't something labeled "Advanced" typically considered "Non-essential"? It would help users if each category had a one-line description clarifying its meaning. I suggest adding a brief introduction under the heading `## Concepts List` with these definitions.

heysokam commented:
> 1. The `partials` imported into the Roadmap from the Creators and VisualScript sites use `title`s that merely indicate a relative path, such as **`assets/cms/panel/description`** from Creators, in contrast to TypeScript which uses a descriptive custom `title` like `Logic vs Data`.

Every roadmap is marked as `draft: true`, because they are not ready to be presented.
This PR doesn't add their content, just their skeleton implementation for future work.
Their title/content is a hint to the writer on where exactly that concept is defined, so they can access the file and modify their temporary dummy/temporary data.

> 2. It appears that these concepts are intended to be categorized, as indicated by labels such as "Advanced Non-essentials", "Mastery", etc. Is there a directory organization within the system that reflects these categories?
`Essentials/Non-Essentials` is equivalent to `Core/Advanced`.
As it was brought on before, `Core/Advanced` could be confusing in their meaning, so they are clarified with -both-, even if they are internally redundant, to clarify this potential ambiguity.
They cannot be separated by `Core/Advanced` folders, because there are meta-level categories inside the more important conceptual overview of the engine (which is already complex enough).

As a concrete example, the `Assets` folder contains concepts that fit all three categories, so subdividing by that meta-category would triplicate the amount of places where the `Asset`-related concepts would be stored.

> 3. Regarding the categories mentioned in point 2, it could be beneficial to define what each category encompasses. For example, the term "Advanced Non-essentials" could be confusing— isn't something labeled "Advanced" typically considered "Non-essential"? It would help users if each category had a one-line description clarifying its meaning. I suggest adding a brief introduction under the heading `## Concepts List` with these definitions.

We could clarify their meaning in the navigation guide that is pending to be written in the `Overview` page for each microsite.

heysokam commented:
> @heysokam, regarding headers in partials rendering in MDX within the table of contents, Docusaurus has included the [MDX partials table-of-contents](https://docusaurus.io/blog/releases/3.2#mdx-partials-table-of-contents) feature as of version 3.2. If we upgrade Docusaurus to the latest version, we can significantly improve TOC navigation.

I didn't know about this feature. Thanks for letting me know. Will look into it :writing_hand:

> An extensive list with direct links and enhanced navigation within the roadmap would greatly boost readability and usability. How feasible is it to update to the latest version of Docusaurus? I recall we already have an issue logged for this update.

`v3.2.0` broke all of our CSS on Release builds.
I haven't tested `v3.2.1` yet, and I don't know if they applied the tailwind fixes from Canary into that version. Their fixes were implemented only halfway through last week.
I will have to research to be able to answer the "how feasible" question. Will let you know.

heysokam commented:
@armando-salazar Apparently the _Markdown Partials Titles_ feature is more of a "bug-fix" to something that one would expect it would work out of the box (I thought it did work, and didn't even notice that it doesn't until I opened that article).

As explained in the confluence page I linked, partials (as a general rule) should not contain any titles (unless explicitly required by the context in which they are on, and they are not used in any other way anywhere else).

This is not because Markdown Partials do not support titles in Docusaurus, but because introducing a `##` title in a programmatical way means that **every** usage of that partial **must** be inserted under a `#` title **mandatory**, or else the TOC of the target page would breal.

The reason for that decision is not technical, but because of the design/architecture of the site.
Using titles in such a way would restrict the use-cases of reusable content.
I'm certain there are uses for such a type of partial (in fact I have a few in this PR, iirc), but they are more the exception than the norm.

armando-salazar commented:
> @armando-salazar Apparently the _Markdown Partials Titles_ feature is more of a "bug-fix" to something that one would expect it would work out of the box (I thought it did work, and didn't even notice that it doesn't until I opened that article).
> 
> As explained in the confluence page I linked, partials (as a general rule) should not contain any titles (unless explicitly required by the context in which they are on, and they are not used in any other way anywhere else).
> 
> This is not because Markdown Partials do not support titles in Docusaurus, but because introducing a `##` title in a programmatical way means that **every** usage of that partial **must** be inserted under a `#` title **mandatory**, or else the TOC of the target page would break.
> 
> The reason for that decision/rule/guideline is not a technical limitation, but because of the design/architecture of the site. Using titles in such a way would restrict the use-cases of reusable content. I'm certain there are uses for such a type of partial _(in fact I have a few in this PR, iirc)_, but they are more the exception than the norm.

I understand your concern. The `###` can indeed cause issues if not nested under a `#`, which is also enforced by Vale rules. Every document written in markdown should start with a level one header (`#`) as the document's title. This is a fundamental guideline for markdown documentation. Could you please provide an example where a partial would be inserted into a page that doesn't already include a `#`? This would help us understand the scenario better and address the issue effectively.

heysokam commented:
> I understand your point. If consistency is our goal, perhaps the TypeScript site should also use relative paths to guide content modification, similar to other sections. For instance, the location of the **Inverse Kinematics** content in TypeScript isn't clear. To maintain consistency and improve file visibility across all sites, I suggest we continue using relative paths for now and consider custom `titles`later.

There wasn't any "consistency goal" really. It was more of an implementation accident.
I implemented the Typescript site manually, and it took too many hours. So I switched my approach and implemented the visualscript+creator folder programmatically with a script that took the folder path and added it to the title+description.

I can probably add the folder to the description of the typescript concepts, in order to make them easier to find by whoever writes their content. Changing their title would be losing work, but changing their description would be very easy, as they all contain the exact same description.
Will switch the PR to draft, add this as a task and reopen when its done. :writing_hand:

> It appears that the category labels don’t impact the directory organization for partials. However, we seem to be using multiple terms interchangeably in our documentation that essentially mean the same thing:
> 
> * **Core** = **Essential** = **Core Essential**
> * **Advanced** = **Non-essentials** = **Advanced Non-essential**
> 
> Could we standardize on a single term for each concept across all documentation? This is crucial to avoid confusion and ensure clarity as we proceed with implementations. Standardization in naming, especially in written content, is essential.

I personally believe this is intuitive, and clearly expresses the multi-dimension of the concept/idea, since it represents both their "relevance" and their "mandatory-ness", all in one combined term. I personally can't see how this could create confusion.
That's why this exact word choice was picked.

But, that said, please let me know which term I should pick and I will modify it all to only use whatever term you decide.

> While I support the idea of defining the content structure on the **Overview** page to guide users, I question the need to clarify the labels from the **Roadmap** page there. These labels are directly presented on the **Roadmap** page; hence, defining them in the titles there would be more beneficial for immediate understanding. Placing this information on a separate page might not serve the user well since our documentation is not typically consumed like a book.

It could very well be in both, using a Partial, because the Concept component is intended for use all throughout the entire website.

There is certainly a bigger need for it in the Roadmaps, due to the current WIP state of of the website.
But, given the user flow of the website, this is something that the user is going to find everywhere. And they will first find the overview and then the concepts component, before they reach the Roadmap at all.
Which makes me think that it should probably be in both, not either/or :thinking:

heysokam commented:
> I understand your concern. The `###` can indeed cause issues if not nested under a `#`, which is also enforced by Vale rules. Every document written in markdown should start with a level one header (`#`) as the document's title. This is a fundamental guideline for markdown documentation. Could you please provide an example where a partial would be inserted into a page that doesn't already include a `#`? This would help us understand the scenario better and address the issue effectively.

I meant in terms of adding a Partial that contains a `### Title` right under a `#### Title`, or similarly incorrect behavior

A potentially conflicting Markdown Partial:
```md
### This is a level3 Title
Some text content
```

An MDX file where the above partial is inserted into multiple wrong places:
```ts
import Thing from '@site/docs/_partials/thething.md

# Title of the Page
## Subtitle
Here our component won't break anything, because it is nested inside a `##`
<Thing />

### Third Level Subtitle
I want to add my Thing here, nested under this subtitle.
But it will not nest it, it will create another Third Level Subtitle and break my intended ToC.
<Thing />

#### Fourth Level
Like above, this won't nest my content. The Partial will create another Third Level subtitle, and exit my currently intended `###`, because my `Thing` Partial contains a `###`.
This means that I'm not allowed to reuse its content where its `#` level doesn't fit, reducing its usability, use-cases, and potentially introducing a ToC "bug" when misused.
```

armando-salazar commented:
> > I understand your point. If consistency is our goal, perhaps the TypeScript site should also use relative paths to guide content modification, similar to other sections. For instance, the location of the **Inverse Kinematics** content in TypeScript isn't clear. To maintain consistency and improve file visibility across all sites, I suggest we continue using relative paths for now and consider custom `titles`later.
> 
> There wasn't any "consistency goal" really. It was more of an implementation accident. I implemented the Typescript site manually, and it took too many hours. So I switched my approach and implemented the visualscript+creator folder programmatically with a script that took the folder path and added it to the title+description.
> 
> I can probably add the folder to the description of the typescript concepts, in order to make them easier to find by whoever writes their content. Changing their title would be losing work, but changing their description would be very easy, as they all contain the exact same description. Will switch the PR to draft, add this as a task and reopen when its done. ✍️
> 

Sounds good to me. Thanks!

> > It appears that the category labels don't impact the directory organization for partials. However, we seem to be using multiple terms interchangeably in our documentation that essentially mean the same thing:
> > 
> > * **Core** = **Essential** = **Core Essential**
> > * **Advanced** = **Non-essentials** = **Advanced Non-essential**
> > 
> > Could we standardize on a single term for each concept across all documentation? This is crucial to avoid confusion and ensure clarity as we proceed with implementations. Standardization in naming, especially in written content, is essential.
> 
> I personally believe this is intuitive, and clearly expresses the multi-dimension of the concept/idea, since it represents both their "relevance" and their "mandatory-ness", all in one combined term. I personally can't see how this could create confusion. That's why this exact word choice was picked.
> 
> But, that said, please let me know which term I should pick and I will modify it all to only use whatever term you decide.
> 

Let's prioritize clarity and simplicity. My suggestion, assuming there are no mandatory terms we must use, we can go with:

- Essentials
- Advanced
- Mastery

When we decide to change these terms, it's important to update all other sources where the term "Core" is mentioned. This includes directory names, notes, docs, etc.

> > While I support the idea of defining the content structure on the **Overview** page to guide users, I question the need to clarify the labels from the **Roadmap** page there. These labels are directly presented on the **Roadmap** page; hence, defining them in the titles there would be more beneficial for immediate understanding. Placing this information on a separate page might not serve the user well since our documentation is not typically consumed like a book.
> 
> It could very well be in both, using a Partial, because the Concept component is intended for use all throughout the entire website.
> 
> There is certainly a bigger need for it in the Roadmaps, due to the current WIP state of of the website. But, given the user flow of the website, this is something that the user is going to find everywhere. And they will first find the overview and then the concepts component, before they reach the Roadmap at all. Which makes me think that it should probably be in both, not either/or 🤔

Excellent point. It would be beneficial to define it on both pages. We should aim for clear and concise language to avoid over-explaining. Here's a suggested format:

"The iR Engine's concepts are organized into levels of increasing complexity to guide your learning progression. Below is a breakdown to help you better understand each level:

- **Essentials**: Fundamental concepts of the engine provide the essential knowledge required to operate the engine effectively.
- **Advanced**: Explore the engine's deeper functionalities and more complex features, enhancing your ability to tackle challenging tasks.
- **Mastery:** Specialized knowledge and skills that enable expert use and mastery of the engine’s full capabilities."

armando-salazar commented:
Thank you for the detailed example. Can we discuss an alternative to enhance the usability and readability of the Learning Roadmap? While the partials are excellent resources, I believe this might be an overengineering effort that could deteriorate the reader's experience while navigating through our docs. 

Let's discuss this offline and explore potential improvements for the navigation.

armando-salazar commented:
> @armando-salazar
> 
> > I understand your concern. The `###` can indeed cause issues if not nested under a `#`, which is also enforced by Vale rules. Every document written in markdown should start with a level one header (`#`) as the document's title. This is a fundamental guideline for markdown documentation. Could you please provide an example where a partial would be inserted into a page that doesn't already include a `#`? This would help us understand the scenario better and address the issue effectively.
> 
> I meant in terms of adding a Partial that contains a `### Title` right under a `#### Title`, or similarly incorrect behavior
> 
> A potentially conflicting Markdown Partial:
> 
> ```
> ### This is a level3 Title
> Some text content
> ```
> 
> An MDX file where the above partial is inserted into multiple wrong places:
> 
> ```
> import Thing from '@site/docs/_partials/thething.md
> 
> # Title of the Page
> ## Subtitle
> Here our component/partial won't break anything, because it is nested inside a `##`
> <Thing />
> 
> ### Third Level Subtitle
> I want to add my Thing here, nested under this subtitle.
> But it will not nest it, it will create another Third Level Subtitle and break my intended ToC.
> <Thing />
> 
> #### Fourth Level
> <Thing />
> Like above, this won't nest my content.
> The Partial will create another Third Level subtitle, and exit my currently intended `####`,
> because my `Thing` Partial contains a `###`.
> This means that I'm not allowed to reuse its content where its `#` level doesn't fit,
> reducing its usability, use-cases, and potentially introducing a ToC "bug" when misused.
> ```
> 
> > **Edit:**
> > I edited the confluence page to also contain a copy of this example, and clarify how/why titles can be conflictive when used within partials.

Thank you for the detailed example. Can we discuss an alternative to enhance the usability and readability of the Learning Roadmap? While the partials are excellent resources, I believe this might be an overengineering effort that could deteriorate the reader's experience while navigating through our docs. 

Let's discuss this offline and explore potential improvements for the navigation.
